mongo/collection.go

```diff
@@ -1174,6 +1189,14 @@ func (coll *Collection) Find(ctx context.Context, filter interface{},
 		}
 		op.Hint(hint)
 	}
+	if fo.Let != nil {
+		let, err := transformBsoncoreDocument(coll.registry, fo.Let, true, "let")
+		if err != nil {
+			closeImplicitSession(sess)
+			return nil, err
```

Comment:
Arguably out-of-scope for this PR, but a more reliable pattern to ensure any created implicit sessions are closed if `Find` returns an error is to used named returns and defer a block that checks for errors and closes the implicit session if one was created.

E.g.
```go
func (coll *Collection) Find(ctx context.Context, filter interface{},
	opts ...*options.FindOptions) (cur *Cursor, err error) {
	// ...
	sess := sessionFromContext(ctx)
	// Always close any created implicit sessions if Find returns an error.
	defer func() {
		if err != nil && sess != nil {
			closeImplicitSession(sess)
		}
	}()
	// ...
```

Reply:
Ah interesting. Is it a common pattern to check a named error return with a `defer`? I don't see that in other places in the driver. I was under the impression that named return values could cause ambiguity, but maybe I'm just thinking about naked returns.

Reply:
It's not a common pattern in the driver right now, but it is used in [pool.go](https://github.com/mongodb/mongo-go-driver/blob/c999a05f708a268ceab67581405205b02485d9d3/x/mongo/driver/topology/pool.go#L293) and in the Go stdlibs [in some places](https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/net/http/transport.go;l=1328) (it's also used without a clear purpose in some [other places](https://github.com/mongodb/mongo-go-driver/blob/c999a05f708a268ceab67581405205b02485d9d3/bson/bsonrw/extjson_reader.go#L272) in the driver). You're right that named return values can cause code to be hard to read when combined with naked returns. However, if we don't use naked returns and only use named return values, the readability and semantics of the code doesn't change, you just guarantee that the named return values are always available inside the function scope.

It's still best practice to avoid named return values because it enables problematic code patterns, but it could be a reasonable tradeoff here since the alternative is possibly forgetting to close an implicit session.

Reply:
Gotcha. Sounds like this is a good use-case for the named return values (and we're not doing a naked return). I added in the `defer` you wrote above for both `Find` and `aggregate`. Seems better to just do this now if it's an improvement rather than file a ticket that may be backlogged. Let me know if it looks good.